tokio/tests/io_async_fd.rs

```diff
@@ -0,0 +1,279 @@
+#![warn(rust_2018_idioms)]
+#![cfg(all(
+    unix,
+    feature = "async-fd",
+    feature = "test-util",
+    feature = "macros",
+    feature = "rt-threaded"
+))]
+
+use futures::FutureExt;
+use std::io::{self, ErrorKind, Read, Write};
+use std::{os::unix::io::RawFd, sync::Arc, time::Duration};
+use tokio::io::AsyncFd;
+
+use nix::errno::Errno;
+use nix::unistd::{close, read, write};
+
+fn is_blocking(e: &nix::Error) -> bool {
+    Some(Errno::EAGAIN) == e.as_errno()
+}
+
+struct FileDescriptor {
+    fd: RawFd,
+}
+
+impl Read for FileDescriptor {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match read(self.fd, buf) {
+            Ok(n) => Ok(n),
+            Err(e) if is_blocking(&e) => Err(ErrorKind::WouldBlock.into()),
+            Err(e) => Err(io::Error::new(ErrorKind::Other, e)),
+        }
+    }
+}
+
+impl Write for FileDescriptor {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        match write(self.fd, buf) {
+            Ok(n) => Ok(n),
+            Err(e) if is_blocking(&e) => Err(ErrorKind::WouldBlock.into()),
+            Err(e) => Err(io::Error::new(ErrorKind::Other, e)),
+        }
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl Drop for FileDescriptor {
+    fn drop(&mut self) {
+        let _ = close(self.fd);
```

Comment:
It is not a good idea to close a file descriptor that is not opened by `AsyncFd`. Consider that other types are initialized by `from_raw_fd`, closing the `fd` here has side effects on those instances.